domains/client/domain-operator/src/domain_bundle_proposer.rs

```diff
@@ -109,74 +109,81 @@ where
+                    .is_within_tx_range(parent_hash, pending_tx_data, &bundle_vrf_hash, &tx_range)
+                    .map_err(|err| {
+                        tracing::error!(
+                            ?err,
+                            ?pending_tx_data,
+                            "Error occurred in locating the tx range"
+                        );
+                    })
+                    .unwrap_or(false);
+                if !is_within_tx_range {
+                    continue;
+                }
 
-            let next_bundle_size = bundle_size + pending_tx_data.encoded_size() as u32;
-            if next_bundle_size > domain_block_limit.max_block_size {
-                break;
-            }
+                let tx_weight = runtime_api_instance
+                    .extrinsic_weight(parent_hash, pending_tx_data)
+                    .map_err(|error| {
+                        sp_blockchain::Error::Application(Box::from(format!(
+                            "Error getting extrinsic weight: {error}"
+                        )))
+                    })?;
+                let next_estimated_bundle_weight =
+                    estimated_bundle_weight.saturating_add(tx_weight);
+                if next_estimated_bundle_weight.any_gt(domain_block_limit.max_block_weight) {
+                    break;
+                }
 
-            estimated_bundle_weight = next_estimated_bundle_weight;
-            bundle_size = next_bundle_size;
-            extrinsics.push(pending_tx_data.clone());
+                let next_bundle_size = bundle_size + pending_tx_data.encoded_size() as u32;
+                if next_bundle_size > domain_block_limit.max_block_size {
+                    break;
+                }
+
+                estimated_bundle_weight = next_estimated_bundle_weight;
+                bundle_size = next_bundle_size;
+
+                // Double check the transaction validity, because the tx pool are re-validate the transaction
+                // in pool asynchronously so there is race condition that the operator imported a domain block
+                // and start producing bundle immediately before the re-validation based on the latest block
+                // is finished, cause the bundle contains illegal tx accidentally and being considered as invalid
+                // bundle and slashing on the honest operator.
+                //
+                // This check is done in similar fashion to block builder's build block.
+                // This check needs to be the last check as otherwise if the tx won't be part of bundle due to
+                // some other checks, its side effect will still be part of RuntimeApiImpl's changes buffer.
+                let transaction_validity_result =
+                    runtime_api_instance.execute_in_transaction(|api| {
+                        let transaction_validity_result = api
+                            .check_transaction_validity_and_do_pre_dispatch(
```

Comment:
If we are doing this check on bundle level, should we do this check on individual extrinsics above?
Would they go through the same approach making individual extrinsic check redundant ?

Reply:
We don't need the individual extrinsic level check IMHO since the bundle level is a superset of it. Ideally, we should perform the check on the block level (like the consensus chain does when producing block) but bundle level is the best we can do.